### ISSUE TITLE
`init_printing()` resulting infinite loop when printing in LaTeX image

### ISSUE THREAD

**Issue by JSS95** [sympy/sympy#18155]
This symptom occurs only in IDE environment such as Spyder or Atom, where the formula image is rendered by latex.

When I start using sympy with `from sympy import *`, everything's just fine. Formulae are shown nicely.

But when I run `init_printing()`, formulae images are not displayed but results an infinite loop.

I believe something is going wrong when sympy tries to use the LaTeX program installed in my computer because:

1. When I do the same job on my labtob, where LaTeX is not installed, nothing goes wrong.

2. If I pass 'mathjax', 'matplotlib', or 'svg' to `use_latex` parameter of `init_printing()`, this symtom does not happen. It only happens when I pass 'png' to it, or when I pass no parameter at all.

**Comment by oscarbenjamin** [comment]
Which version of SymPy are you using?

**Comment by JSS95** [comment]
> Which version of SymPy are you using?

It's 1.5 release version, and same issue happens in latest development version as well.

**Comment by oscarbenjamin** [comment]
Have you tried version 1.4?

**Comment by JSS95** [comment]
> Have you tried version 1.4?

Yes. In fact, this was happening before 1.5 was released. I have forgotten to report it until now.

**Comment by WenyinWei** [comment]
The same with me, I have gotten stuck in the infinite wait of expression output in jupyter notebook for about half a year on Windows 10, no matter which version of python it is, 37, 38 or 39. The bug only happens on one of my laptops so I can switch to another Windows machine to avoid it. While on Ubuntu, everything goes well even when the OS is running as a WSL container and porting its underlying jupyter to the Windows system which itself fails to print the formula and just keep kernel running.

**Comment by asmeurer** [comment]
Are you able to interrupt the kernel and get a traceback of where it is stuck that you can paste here?

**Comment by WenyinWei** [comment]
Sorry, I tried `trace_set()` according to [What is the right way to debug in iPython notebook?](https://stackoverflow.com/questions/32409629/what-is-the-right-way-to-debug-in-ipython-notebook), but it can only set breakpoints and check variables rather than telling me in which function it is stuck. Could you please tell me how to debug then?

**Comment by ChrisChemHater** [comment]
Same problem for sympy 1.9 and python 3.9.4. I've tried run jupyter notebook on vscode and html, they all stucked when printing expressions.

**Comment by idanpa** [comment]
The hang in my case (Windows 10, sympy 1.9) was an stdout.read() hang after calling dvipng.exe ([sympy/printing/preview.py:307](https://github.com/sympy/sympy/blob/master/sympy/printing/preview.py#L307))
I tried to reproduce on terminal and indeed dvipng gets weirdly stuck:
```
PS C:\Users\idanp\AppData\Local\Temp\tmp68e8dbl_> C:\texlive\2020\bin\win32\dvipng.EXE -T tight -z 9 --truecolor -o texput.png texput.dvi
This is C:\texlive\2020\bin\win32\dvipng.EXE 1.17 Copyright 2002-2015, 2019 Jan-Ake Larsson
[1 <raw PostScript
``` 
Setting the debug arg (`-d`) on dvipng solves this issue, png is generated and dvipng exits successfully 🤷‍♂️.
Same behavior on the other dvi converters too.
So my workaround is to add `cmd.extend(["-d"])` on preview.py (but need to understand what is wrong with dvipng.exe on Windows...).

**Comment by oscarbenjamin** [comment]
I think that the reason this has gone unfixed for so long is because most people are unable to reproduce it which is in turn because it is some Windows-specific bug either in dvipng or in the way that sympy calls dvipng. Ultimately this can only really be fixed by someone who can actually reproduce it.

It is possible that the hang is to do with the way that sympy calls the subprocess. If you can figure out what input is being passed to dvipng and see if it works when run directly on the command line then it might be that the bug is in sympy and there is nothing actually wrong with dvipng. The code in sympy for doing this is old and the subprocess module has newer high-level interfaces for doing things like this. It is possible that the sympy code should be changed to use `communicate` or `run` or something else rather than `.read()`.

**Comment by idanpa** [comment]
@oscarbenjamin - this is not about how sympy calls dvipng, this is real hang that I can reproduce on the terminal, e.g. take the following latex file (texput.tex):

``` latex
\documentclass[varwidth,10pt]{standalone}
\usepackage{amsmath}
\usepackage{amsfonts}
\begin{document}
$3$
\end{document}
```

Generate .dvi with:
```
latex -halt-on-error -interaction=nonstopmode texput.tex
```
and then
```
dvipng texput.png texput.dvi
```
Hangs with:
```
This is C:\texlive\2024\bin\windows\dvipng.exe 1.17 Copyright 2002-2015, 2019 Jan-Ake Larsson
[1 <raw PostScript
```

Seems like a bug in dvipng, it is get resolved if I use `\documentclass[preview]{standalone}` instead of `varwidth`.

Solutions I see:
1. replace `varwidth` to `preview` (this would remove the control on the font size and would use the default size)
2. do what IPython's latex_to_png is doing, use `\documentclass{article}`  and `\pagestyle{empty}`
3. call IPython's latex_to_png instead of implementing in sympy
4. remove all latex-to-picture code from sympy, let the kernel clients to handle the png generation (like currently being done in common jupyter clients when they get `text/latex` payload)


Please let me know what you think and I can send a PR for review.

**Comment by oscarbenjamin** [comment]
My preference is option 4: remove all latex to picture code from sympy. At least we should not use sympy's latex to picture code by default from things like `init_printing()`.

There is a good chance that others would disagree though. I don't have enough experience of using these different UIs and their backend interfaces to know what the implications would be.

**Comment by asmeurer** [comment]
Keeping the code for this in the Jupiter kernels instead of sympy maybe makes sense in principle but I don't know how well these things get maintained when they live in IPython vs. in sympy. Also converting an equation to an image is independent of using Jupiter or a notebook and people use preview() for those other purposes too.

**Comment by idanpa** [comment]
@asmeurer,
What do you think about making `use_latex` to be just True/False property, where true only set 
```
Printable._repr_latex_ = _print_latex_text
```
which seems like the sane thing to do in general.
Now this happens only when `use_latex in (True, 'mathjax')` (why mathjax related to this?))

BTW - on first glance IPython has more extensive tests for the latex_to_png flow.

**Comment by idanpa** [comment]
I think the default .init_printing() should just work out of the box on windows.
It seems like the workaround #27415 is not accepted.

Trying to understand the real problem, does anyone have a windows machine with latex where .init_printing() works on qtconsole? What is the setup?

I can tell that this bug happens for me on both windows 10 and windows 11 machines with dvipng version 1.17 from either MikTex or texlive. 

I can add the debug flag to dvipng call if on windows (if you think this is reasonable) as this solves the problem too.

**Comment by oscarbenjamin** [comment]
Has this bug been reported to dvipng somewhere?

**Comment by idanpa** [comment]
Reported now https://lists.nongnu.org/archive/html/dvipng/2025-01/msg00000.html

**Comment by moorepants** [comment]
The thread on the dvipng report seems to indicate: "adding --norawps solves the problem.", so maybe we can add that flag to our call of dvipng.
